graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/util/ZipUtils.java

```diff
@@ -0,0 +1,188 @@
+     * method to zip a file / several files / folder or add files to an existing zip file (without duplicates)
+     *
+     * @param nodes list of nodes to be included in the new zip file (only files or folders)
+     * @param file  the zip file
+     */
+    public static void addToZip(List<JCRNodeWrapper> nodes, JCRNodeWrapper file) {
+        InputStream is = null;
+        String filename = file.getName();
+        File tmp = null;
+        try {
+            tmp = File.createTempFile(filename, "");
+            ZipOutputStream zout = new ZipOutputStream(new FileOutputStream(tmp));
+
+            //if the zip file already exists we transfer all its subfiles into the new zip file
+            if (file.hasNode(Constants.JCR_CONTENT) && file.getNode(Constants.JCR_CONTENT).hasProperty(Constants.JCR_DATA)) {
+                ZipInputStream zin = new ZipInputStream(file.getFileContent().downloadFile());
+                ZipEntry entry = zin.getNextEntry();
+                while (entry != null) {
+                    zout.putNextEntry(entry);
+                    copy(zin, zout);
+                    entry = zin.getNextEntry();
+                }
+                zin.close();
+                //else, we add mandatory child nodes
+            } else {
+                file.addNode(Constants.JCR_CONTENT, Constants.JAHIANT_RESOURCE);
+            }
+            //now we add the new files/directories
+            for (JCRNodeWrapper node : nodes) {
+                zip(node, zout, "");
+            }
+            zout.close();
+            is = new BufferedInputStream(new FileInputStream(tmp));
+            //put the new zip file into jcr:data
+            file.getNode(Constants.JCR_CONTENT).setProperty(Constants.JCR_DATA, new BinaryImpl(is));
+            file.getNode(Constants.JCR_CONTENT).setProperty(Constants.JCR_MIMETYPE, "application/zip");
+        } catch (IOException | RepositoryException e) {
+            throw new DataFetchingException(e);
+        } finally {
+            IOUtils.closeQuietly(is);
+            FileUtils.deleteQuietly(tmp);
+        }
+    }
+
+    /**
+     * method to unzip a zip file with all its tree
+     *
+     * @param dest    destination node (folder)
+     * @param zipFile zip file to unzip
+     */
+    public static void unzip(JCRNodeWrapper dest, JCRNodeWrapper zipFile) {
+        try {
+            File tmp;
+            ZipInputStream zis = new ZipInputStream(zipFile.getFileContent().downloadFile());
+            ZipEntry entry = zis.getNextEntry();
+            while (entry != null) {
+
+                if (entry.isDirectory()) {
+                    //if the entry is a directory, create it to build the whole tree
+                    dest.addNode(entry.getName(), Constants.JAHIANT_FOLDER);
+
+                } else {
+
+                    JCRNodeWrapper file = dest.addNode(entry.getName(), Constants.JAHIANT_FILE);
+                    file.addNode(Constants.JCR_CONTENT, Constants.JAHIANT_RESOURCE);
```

Comment:
You can use file.uploadFile(is, mimetype) - it creates the jcr:content, set the data and the mimetype in one call